network/p2p/unicast/manager.go

```diff
@@ -0,0 +1,175 @@
+package unicast
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	libp2pnet "github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// MaxConnectAttemptSleepDuration is the maximum number of milliseconds to wait between attempts for a 1-1 direct connection
+const MaxConnectAttemptSleepDuration = 5
+
+// Manager manages libp2p stream negotiation and creation, which is utilized for unicast dispatches.
+type Manager struct {
+	logger         zerolog.Logger
+	streamFactory  StreamFactory
+	unicasts       []Protocol
+	defaultHandler libp2pnet.StreamHandler
+	sporkId        flow.Identifier
+}
+
+func NewUnicastManager(logger zerolog.Logger, streamFactory StreamFactory, sporkId flow.Identifier) *Manager {
+	return &Manager{
+		logger:        logger,
+		streamFactory: streamFactory,
+		sporkId:       sporkId,
+	}
+}
+
+// WithDefaultHandler sets the default stream handler for this unicast manager. The default handler is utilized
+// as the core handler for other unicast protocols, e.g., compressions.
+func (m *Manager) WithDefaultHandler(defaultHandler libp2pnet.StreamHandler) {
+	defaultProtocolID := FlowProtocolID(m.sporkId)
+	m.defaultHandler = defaultHandler
+
+	if len(m.unicasts) > 0 {
+		panic("default handler must be set only once before any unicast registration")
+	}
+
+	m.unicasts = []Protocol{
+		&PlainStream{
+			protocolId: defaultProtocolID,
+			handler:    defaultHandler,
+		},
+	}
+
+	m.streamFactory.SetStreamHandler(defaultProtocolID, defaultHandler)
+}
+
+// Register registers given protocol name as preferred unicast. Each invocation of register prioritizes the current protocol
+// over previously registered ones.
+func (m *Manager) Register(unicast ProtocolName) error {
+	factory, err := ToProtocolFactory(unicast)
+	if err != nil {
+		return fmt.Errorf("could not translate protocol name into factory: %w", err)
+	}
+
+	u := factory(m.logger, m.sporkId, m.defaultHandler)
+
+	m.unicasts = append(m.unicasts, u)
+	m.streamFactory.SetStreamHandler(u.ProtocolId(), u.Handler())
+
+	return nil
+}
+
+// CreateStream tries establishing a libp2p stream to the remote peer id. It tries creating streams in the descending order of preference until
+// it either creates a successful stream or runs out of options. Creating stream on each protocol is tried at most `maxAttempt` one, and then falls
+// back to the less preferred one.
+func (m *Manager) CreateStream(ctx context.Context, peerID peer.ID, maxAttempts int) (libp2pnet.Stream, []multiaddr.Multiaddr, error) {
+	var errs error
+
+	for i := len(m.unicasts) - 1; i >= 0; i-- {
+		s, addrs, err := m.createStreamWithProtocol(ctx, m.unicasts[i].ProtocolId(), peerID, maxAttempts)
+		if err != nil {
```

Comment:
we will need to put in a check here to bail out and return if the stream connection fails because of any reason other than "protocol not supported" else we will end up making an extra attempt.
May be address that as a separate issue since you will have to define new error types to check against.

Reply:
Thanks for the comment, issued up here: https://github.com/dapperlabs/flow-go/issues/6036